File: frontend/src/app/shared/validators/cpf.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * CPF Validator - Validates Brazilian CPF with check digits
 * Story 3.1: Supplier Management
 */
export function cpfValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const cpf = control.value;

    if (!cpf) {
      return null; // Don't validate empty values (use Validators.required for that)
    }

    // Remove formatting
    const cpfClean = cpf.replace(/\D/g, '');

    // Check length
    if (cpfClean.length !== 11) {
      return { cpf: { message: 'CPF deve ter 11 dígitos' } };
    }

    // Check if all digits are the same
    if (/^(\d)\1{10}$/.test(cpfClean)) {
      return { cpf: { message: 'CPF inválido' } };
    }

    // Validate check digits
    if (!isValidCpf(cpfClean)) {
      return { cpf: { message: 'CPF inválido' } };
    }

    return null;
  };
}

function isValidCpf(cpf: string): boolean {
  try {
    // First check digit
    let sum = 0;
    for (let i = 0; i < 9; i++) {
      sum += parseInt(cpf.charAt(i)) * (10 - i);
    }
    let digit1 = 11 - (sum % 11);
    digit1 = (digit1 >= 10) ? 0 : digit1;

    if (parseInt(cpf.charAt(9)) !== digit1) {
      return false;
    }

    // Second check digit
    sum = 0;
    for (let i = 0; i < 10; i++) {
      sum += parseInt(cpf.charAt(i)) * (11 - i);
    }
    let digit2 = 11 - (sum % 11);
    digit2 = (digit2 >= 10) ? 0 : digit2;

    return parseInt(cpf.charAt(10)) === digit2;

  } catch (e) {
    return false;
  }
}

/**
 * Format CPF to standard format: 000.000.000-00
 */
export function formatCpf(cpf: string): string {
  if (!cpf) {
    return '';
  }

  const cpfClean = cpf.replace(/\D/g, '');

  if (cpfClean.length !== 11) {
    return cpf;
  }

  return `${cpfClean.substring(0, 3)}.${cpfClean.substring(3, 6)}.${cpfClean.substring(6, 9)}-${cpfClean.substring(9)}`;
}

/**
 * Clean CPF formatting
 */
export function cleanCpf(cpf: string): string {
  return cpf ? cpf.replace(/\D/g, '') : '';
}

File: backend/src/main/java/com/estoquecentral/shared/validator/CpfValidator.java
package com.estoquecentral.shared.validator;

/**
 * CPF (Cadastro de Pessoas Físicas) validator for Brazilian tax IDs.
 *
 * Validates CPF format and check digits according to Brazilian government rules.
 *
 * Story 4.1: Customer Management - AC4 (CPF validation)
 */
public class CpfValidator {

    private static final int CPF_LENGTH = 11;

    /**
     * Validates a CPF string.
     *
     * @param cpf the CPF to validate (with or without formatting)
     * @return true if CPF is valid, false otherwise
     */
    public static boolean isValid(String cpf) {
        if (cpf == null || cpf.isEmpty()) {
            return false;
        }

        // Remove formatting characters
        cpf = cpf.replaceAll("[^0-9]", "");

        // Check length
        if (cpf.length() != CPF_LENGTH) {
            return false;
        }

        // Check for known invalid CPFs (all digits the same)
        if (cpf.matches("(\\d)\\1{10}")) {
            return false;
        }

        // Validate first check digit
        int[] multiplicadores1 = {10, 9, 8, 7, 6, 5, 4, 3, 2};
        int soma = 0;
        for (int i = 0; i < 9; i++) {
            soma += Character.getNumericValue(cpf.charAt(i)) * multiplicadores1[i];
        }
        int digito1 = (soma % 11 < 2) ? 0 : 11 - (soma % 11);

        if (Character.getNumericValue(cpf.charAt(9)) != digito1) {
            return false;
        }

        // Validate second check digit
        int[] multiplicadores2 = {11, 10, 9, 8, 7, 6, 5, 4, 3, 2};
        soma = 0;
        for (int i = 0; i < 10; i++) {
            soma += Character.getNumericValue(cpf.charAt(i)) * multiplicadores2[i];
        }
        int digito2 = (soma % 11 < 2) ? 0 : 11 - (soma % 11);

        return Character.getNumericValue(cpf.charAt(10)) == digito2;
    }

    /**
     * Formats a CPF string to the standard format: 000.000.000-00
     *
     * @param cpf the CPF to format (digits only)
     * @return formatted CPF string
     */
    public static String format(String cpf) {
        if (cpf == null || cpf.isEmpty()) {
            return cpf;
        }

        cpf = cpf.replaceAll("[^0-9]", "");

        if (cpf.length() != CPF_LENGTH) {
            return cpf;
        }

        return String.format("%s.%s.%s-%s",
            cpf.substring(0, 3),
            cpf.substring(3, 6),
            cpf.substring(6, 9),
            cpf.substring(9, 11)
        );
    }

    /**
     * Removes formatting from a CPF string (returns digits only).
     *
     * @param cpf the CPF to clean
     * @return CPF with digits only
     */
    public static String cleanFormat(String cpf) {
        if (cpf == null) {
            return null;
        }
        return cpf.replaceAll("[^0-9]", "");
    }
}
